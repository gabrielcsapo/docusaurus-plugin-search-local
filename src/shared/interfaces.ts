import { DocusaurusConfig } from "@docusaurus/types";
import lunr from "lunr";

export interface TranslationMap {
  search_placeholder?: string;
  see_all_results?: string;
  no_results?: string;
  search_results_for?: string;
  search_the_documentation?: string;
  count_documents_found?: string;
  count_documents_found_plural?: string;
  no_documents_were_found?: string;
}

export type SmartTerm = SmartTermItem[];

export interface SmartTermItem {
  value: string;
  trailing?: boolean;
  maybeTyping?: boolean;
}

export interface SmartQuery {
  tokens: string[];
  term: QueryTerm;
}

export type QueryTerm = QueryTermItem[];

export interface QueryTermItem {
  value: string;
  presence: lunr.Query.presence;
  wildcard: lunr.Query.wildcard;
}

export interface WrappedTerm {
  missed: number;
  term: SmartTerm;
}

export type MetadataPosition = [number, number];

export interface MatchMetadata {
  [token: string]: {
    [field: string]: {
      position: MetadataPosition[];
    };
  };
}

export interface HighlightChunk {
  html: string;
  textLength: number;
}

export interface ChunkIndexRef {
  chunkIndex: number;
}

/**
 * properties of document is shorten for smaller serialized search index.
 */
export interface SearchDocument {
  /** Doc ID. */
  i: number;

  /** Doc title. */
  t: string;

  /** Doc URL. */
  u: string;

  /** Doc hash. */
  h?: string;

  /** Doc parent ID. */
  p?: number;

  /** Doc breadcrumb. */
  b?: string[];

  /** Doc section title */
  s?: string;
}

/**
 * - 0: Doc title
 * - 1: Doc heading
 * - 2: Doc content
 */
export type SearchDocumentType = 0 | 1 | 2;

export interface SearchResultBase {
  document: SearchDocument;
  type: SearchDocumentType;
  page: SearchDocument | undefined | false;
  metadata: MatchMetadata;
  tokens: string[];
}

export type SearchResult = SearchResultBase & SearchResultExtra;

export type InitialSearchResult = SearchResultBase & Partial<SearchResultExtra>;

export interface SearchResultExtra {
  score: number;
  index: number;
  isInterOfTree: boolean;
  isLastOfTree: boolean;
}

export interface WrappedIndex {
  documents: SearchDocument[];
  index: lunr.Index;
  type: SearchDocumentType;
}

export interface ParsedDocument {
  pageTitle: string;
  sections: ParsedDocumentSection[];
  breadcrumb: string[];
}

export interface ParsedDocumentSection {
  title: string;
  hash: string;
  content: string;
}

export interface DocInfoWithRoute {
  route: string;
  url: string;
  type: DocInfoType;
}

export interface DocInfoWithFilePath {
  filePath: string;
  url: string;
  type: DocInfoType;
}

export type DocInfoType = "docs" | "blog" | "page";

export interface PluginOptions {
  indexDocs?: boolean;
  indexBlog?: boolean;
  indexPages?: boolean;
  docsRouteBasePath?: string | string[];
  blogRouteBasePath?: string | string[];
  language?: string | string[];
  hashed?: boolean;
  docsDir?: string | string[];
  blogDir?: string | string[];
  removeDefaultStopWordFilter?: boolean;
  highlightSearchTermsOnTargetPage?: boolean;

  searchResultLimits?: number;
  searchResultContextMaxLength?: number;

  translations?: TranslationMap;

  ignoreFiles?: string | RegExp | (string | RegExp)[];
}

export type ProcessedPluginOptions = Required<
  Omit<
    PluginOptions,
    | "language"
    | "docsRouteBasePath"
    | "blogRouteBasePath"
    | "docsDir"
    | "blogDir"
    | "ignoreFiles"
  >
> & {
  docsRouteBasePath: string[];
  blogRouteBasePath: string[];
  language: string[];
  docsDir: string[];
  blogDir: string[];
  ignoreFiles: (string | RegExp)[];
};

export interface PostBuildData {
  routesPaths: string[];
  outDir: string;
  baseUrl: string;
  siteConfig: DocusaurusConfig;
}
