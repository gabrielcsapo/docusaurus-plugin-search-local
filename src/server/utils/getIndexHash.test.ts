import fs from "fs";
import klawSync from "klaw-sync";
import { PluginConfig } from "../../shared/types";
import { getIndexHash } from "./getIndexHash";

jest.mock("klaw-sync");
jest.mock("fs");
const mockConsoleWarn = jest
  .spyOn(console, "warn")
  .mockImplementation(() => void 0);

jest.mock(
  "../../../../package.json",
  () => ({
    version: "0.0.0",
  }),
  {
    virtual: true,
  }
);

(klawSync as jest.MockedFunction<typeof klawSync>).mockImplementation(
  (root, options) => {
    let files: string[] = [];
    if (root === "/tmp/docs") {
      files = ["/tmp/docs/a.md", "/tmp/docs/b.md", "/tmp/docs/b.png"];
    }

    const items = files.map((path) => ({ path } as klawSync.Item));

    return options?.filter ? items.filter(options.filter) : items;
  }
);

(
  fs.readFileSync as jest.MockedFunction<typeof fs.readFileSync>
).mockImplementation((filePath: fs.PathOrFileDescriptor) => {
  if (typeof filePath === "string" && filePath.endsWith(".md")) {
    return Buffer.from(filePath);
  }
  throw new Error(`Unknown file: ${filePath}`);
});

(fs.existsSync as jest.MockedFunction<typeof fs.existsSync>).mockImplementation(
  (filePath: fs.PathOrFileDescriptor) => {
    if (typeof filePath === "string") {
      return filePath.startsWith("/tmp/");
    }

    return false;
  }
);

(fs.lstatSync as jest.MockedFunction<typeof fs.lstatSync>).mockImplementation(
  (filePath: fs.PathOrFileDescriptor) => {
    return {
      isDirectory: () => {
        if (typeof filePath === "string") {
          return !filePath.includes(".");
        }
      },
    } as fs.Stats;
  }
);

describe("getIndexHash", () => {
  test.each<[Partial<PluginConfig>, string | null, number]>([
    [{ hashed: false }, null, 0],
    [{ hashed: true, indexDocs: true, docsDir: ["/tmp/docs"] }, "87def35c", 0],
    [{ hashed: true, indexBlog: true, blogDir: ["/tmp/blog"] }, null, 0],
    [
      { hashed: true, indexDocs: true, docsDir: ["/does-not-exist/docs"] },
      null,
      1,
    ],
    [{ hashed: true, indexDocs: true, docsDir: ["/tmp/index.js"] }, null, 1],
  ])("getIndexHash(%j) should return '%s'", (config, hash, warnCount) => {
    expect(getIndexHash(config as PluginConfig)).toBe(hash);
    expect(mockConsoleWarn).toBeCalledTimes(warnCount);
  });
});
