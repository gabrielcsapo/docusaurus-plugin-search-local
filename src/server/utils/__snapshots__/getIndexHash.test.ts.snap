// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getIndexHash > getIndexHash({"hashed":false}) should return '0' 1`] = `null`;

exports[`getIndexHash > getIndexHash({"hashed":true,"indexBlog":true,"blogDir":["/tmp/blog"]}) should return '0' 1`] = `null`;

exports[`getIndexHash > getIndexHash({"hashed":true,"indexDocs":true,"docsDir":["/does-not-exist/docs"]}) should return '1' 1`] = `null`;

exports[`getIndexHash > getIndexHash({"hashed":true,"indexDocs":true,"docsDir":["/tmp/docs"]}) should return '0' 1`] = `"c9c4419d"`;

exports[`getIndexHash > getIndexHash({"hashed":true,"indexDocs":true,"docsDir":["/tmp/index.js"]}) should return '1' 1`] = `null`;
